### PR TITLE
Adding code to generate ascent backend for Catalyst

### PR DESCRIPTION
**Pull request by ayenpure**
-- adding necessary flags `ENABLE_CATALYST` and `CATALYST_DIR` to
   specify if the backend should be created
-- adding the necessary source code into `src/libs/catalyst` directory

### CODE REVIEW

**Comment by ayenpure** [comment]
@nicolemarsaglia @cyrush 
Can you please review these changes for building Catalyst's ascent backend?

File: src/libs/catalyst/AscentCatalyst.cxx [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+      const auto fname =
+        script.dtype().is_string() ? script.as_string() : script["filename"].as_string();
+      std::cout << " Provided script : " << fname << std::endl;
+      std::ifstream ifs(fname);
+      std::string content;
+      content.assign((std::istreambuf_iterator<char>(ifs)),
+                     (std::istreambuf_iterator<char>()));
+      conduit::Node node = conduit::Node();
+      node.parse(content, "yaml");
+      node.print();
```
Comment (francois-kitware) [inline src/libs/catalyst/AscentCatalyst.cxx]: Is it for debug purpose? Should it be removed in the production version?

File: src/libs/catalyst/AscentCatalyst.cxx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        return instance;
+      }
+    }
+
+    void SetAscent(void* _ascent) { this->ascent = _ascent; }
+    void* GetAscent() { return this->ascent; }
+
+    void AddActions(conduit::Node& node)
+    {
+      actions.push_back(node);
```
Comment (CharlesGueunet) [inline src/libs/catalyst/AscentCatalyst.cxx]:
```suggestion
      this->actions.emplace_back(node);
```

**Comment by cyrush** [comment]
This looks good.

Small suggestion (to follow our undocumented CMake standards for TPLs):

Instead of `ENABLE_CATALYST` + `CATALYST_DIR`, we can look for Catalyst if `CATALYST_DIR` is presented.

Then guard further use of Catalyst with `CATALYST_FOUND` (instead of `ENABLE_CATALYST`)


And a Q: I assume `catalyst_implementation` cmake magic hands all the cmake install targets?

File: src/libs/catalyst/AscentCatalyst.cxx (hunk 1 above)
Comment (ayenpure) [reply src/libs/catalyst/AscentCatalyst.cxx]: Yes, it was intended for debug, I've removed it

File: src/libs/catalyst/AscentCatalyst.cxx (hunk 2 above)
Comment (ayenpure) [reply src/libs/catalyst/AscentCatalyst.cxx]: Thanks for the suggestion; I've updated the code

**Comment by ayenpure** [comment]
> This looks good.
> 
> Small suggestion (to follow our undocumented CMake standards for TPLs):
> 
> Instead of `ENABLE_CATALYST` + `CATALYST_DIR`, we can look for Catalyst if `CATALYST_DIR` is presented.
> 
> Then guard further use of Catalyst with `CATALYST_FOUND` (instead of `ENABLE_CATALYST`)
> 
> And a Q: I assume `catalyst_implementation` cmake magic hands all the cmake install targets?

**Comment by ayenpure** [comment]
@cyrush, I've updated the branch to remote the `ENABLE_CATALYST` flag.
And yes, the `catalyst_implementation` function does the magic for installing the library at proper targets.
Currently, I have it install into `lib/catalyst` in both `build` and `install` directories.
Let me know if you want me to drop `catalyst` and just install it into `lib`

**Comment by cyrush** [comment]
Looks like you were able to reopen, I think we are good.

File: src/libs/CMakeLists.txt [hunk 3]
```diff
@@ -59,3 +59,8 @@ endif()
 # Add ascent
 ################################
 add_subdirectory(ascent)
+
+# Catalyst's ascent implementation
+if(CATALYST_FOUND)
+    add_subdirectory(catalyst)
+endif()
```
Comment (cyrush) [inline src/libs/CMakeLists.txt:66]: Looks good, thanks for these changes!

**Comment by cyrush** [comment]
> @cyrush, I've updated the branch to remote the `ENABLE_CATALYST` flag. And yes, the `catalyst_implementation` function does the magic for installing the library at proper targets. Currently, I have it install into `lib/catalyst` in both `build` and `install` directories. Let me know if you want me to drop `catalyst` and just install it into `lib`

Thanks for confirming about the cmake logic. 

 Installing to `lib` would match all the other libs, so if that's possible, let's aim for that.

After that change I will merge.

**Comment by ayenpure** [comment]
@cyrush, does everything look good to you?

**Review by cyrush — APPROVED** [review]
Looks good, I'll merge pending final CI checks.
Thanks!

**Comment by cyrush** [comment]
@ayenpure merged, thanks for adding this!

**Pull request merged**